src/dataflow/src/source/postgres.rs

```diff
@@ -197,22 +189,76 @@ impl PostgresSourceReader {
                 .await?;
             for msg in data {
                 if let SimpleQueryMessage::Row(row) = msg {
-                    let mut mz_row = Row::default();
-                    let rel_id: Datum = (info.rel_id as i32).into();
-                    mz_row.push(rel_id);
-                    mz_row.push_list((0..row.len()).map(|n| {
-                        let a: Datum = row.get(n).into();
-                        a
-                    }));
+                    let mz_row = try_fatal!(self.row_from_query_row(info.rel_id, row));
                     try_recoverable!(snapshot_tx.insert(mz_row.clone()).await);
                     try_fatal!(buffer.write(&try_fatal!(bincode::serialize(&mz_row))).await);
                 }
             }
         }
         client.simple_query("COMMIT;").await?;
+        for table in publication_tables {
+            self.synced_relids.insert(table.rel_id);
+        }
         Ok(())
     }
 
+    /// Takes a snapshot of the specified relation, syncing the data to the source.
+    /// This function will be called when we detect a new relation has been added to the
+    /// Postgres publication.
+    async fn produce_snapshot(
+        &mut self,
+        rel_id: u32,
+    ) -> Result<(PgLsn, Vec<Row>), ReplicationError> {
+        let config = self.connector.conn.parse()?;
+        let tls = try_recoverable!(postgres_util::make_tls(&config));
+        let (client, connection) =
+            try_recoverable!(tokio_postgres::connect(&self.connector.conn, tls).await);
+        tokio::spawn(connection);
+
+        // Get target table information
+        let tables = try_recoverable!(
+            postgres_util::publication_info(
+                &self.connector.conn,
+                &self.connector.publication,
+                Some(rel_id)
+            )
+            .await
+        );
+        let table = tables.get(0).expect("publication info must exist");
+
+        // Start a transaction and immediately create a replication slot with the USE SNAPSHOT
+        // directive. This makes the starting point of the slot and the snapshot of the transaction
+        // identical.
+        client
+            .simple_query("BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ;")
+            .await?;
+
+        let lsn: PgLsn = client
+            .query_one("SELECT pg_current_wal_lsn()", &[])
```

Comment:
This function unfortunately doesn't return the lsn of this transaction's snapshot, it just tells you what the latest LSN of the server is. I think we need to get the consistent point from the side effect of creating a slot and then immediately discard it after we're done with the transaction